File: algorithms/cpp/serializeandDeserializeBinaryTree/serializeandDeserializeBinaryTree.cpp
// Source: https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
// Author: Miao Zhang
// Date:   2021-01-31

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string res;
        dfs(root, res);
        return res;
    }
    
    void dfs(TreeNode* root, string& res) {
        if (!root) {
            res += "null ";
            return;
        }
        res += to_string(root->val);
        res += " ";
        dfs(root->left, res);
        dfs(root->right, res);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in(data);
        return dfs(in);
    }
    
    TreeNode* dfs(istringstream& in) {
        string val;
        in >> val;
        if (val == "null") return nullptr;
        TreeNode* node = new TreeNode(stoi(val));
        node->left = dfs(in);
        node->right = dfs(in);
        return node;
    }
    
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

File: algorithms/python/serializeandDeserializeBinaryTree/serializeandDeserializeBinaryTree.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
# Author: Miao Zhang
# Date:   2021-01-31

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        res = []
        def dfs(node):
            if not node:
                res.append('null')
                return 
            res.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
    
        dfs(root)
        return ','.join(res)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        def dfs(queue):
            val = queue.pop(0)
            if val == 'null':
                return None
            node = TreeNode(val)
            node.left = dfs(queue)
            node.right = dfs(queue)
            return node

        q = data.split(',')
        return dfs(q)
